Implement port forwarding for windows

benmoss opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds support for port forwarding when running kubelet on Windows nodes

**Which issue(s) this PR fixes**:
Fixes #64531

**Special notes for your reviewer**:
The only way I could figure out to expose access to the container's network interfaces is by running another container in the pod from which we can use a netcat-like program to proxy the TCP stream. The source for this container is [here](https://github.com/benmoss/wincat/blob/master/cmd/wincat/wincat.go).

Right now on Windows nodes it is standard to build the pause image locally, as Microsoft documents in this [script](https://github.com/Microsoft/SDN/blob/81b1ab7cda419ed54a4b36c497b4414bafbf5f91/Kubernetes/windows/InstallImages.ps1). I am thinking that we introduce a similar container that we assume is already on the node in order to avoid pulling this image at runtime and maintaining it in a remote registry.

**Does this PR introduce a user-facing change?**:
```release-note
Kubelet will attempt to use wincat.exe in the pause container for port forwarding when running on Windows
```

Looking for feedback since this is a semi-radical departure from how port-forwarding works on Linux.

/hold
/sig windows
/sig node
cc @yujuhong @feiskyer @astrieanna

benmoss committed:
Implement port forwarding for windows

On Windows the only way to access the container's network interfaces is
by running another process in the pod from which we can use a
netcat-like program to proxy the TCP stream

Proposed wincat.exe can be found here: https://github.com/benmoss/wincat

benmoss commented:
/retest

benmoss commented on pkg/kubelet/dockershim/docker_streaming_windows.go:
```diff
@@ -0,0 +1,95 @@
+	wincat, err := client.CreateContainer(dockertypes.ContainerCreateConfig{
+		Name: name,
+		Config: &dockercontainer.Config{
+			Image:        "gcr.io/cf-london-servces-k8s/windows-images/wincat:latest",
```
Worth pointing out that this specifically shouldn't be merged, my idea is that this would be something like `"wincat"` and would be expected to be on the node already and not pulled from a registry, just as the pause image is today

neolit123 replied:
hm, but isn't the pause image still pulled on Windows nodes if it's not present?
it was recently added to the official MS registry for that purpose.

benmoss replied:
I hadn't heard anything about that. In the [SDN code](https://github.com/Microsoft/SDN/blob/81b1ab7cda419ed54a4b36c497b4414bafbf5f91/Kubernetes/windows/start-kubelet.ps1#L118) the pause container is still being specified as `kubeletwin/pause` which doesn't appear to be on Dockerhub

benmoss replied:
Looks like it's [the same](https://github.com/Azure/aks-engine/blob/09564f98115912a60144771e27b0c1fb2ef3aa77/parts/k8s/windowskubeletfunc.ps1#L133-L157) for aks-engine

neolit123 replied:
this is the image that i saw (+tested):
`mcr.microsoft.com/k8s/core/pause:1.0.0`
(ML 1803/1809, nanoserver based)

https://github.com/kubernetes/kubernetes/issues/74218#issuecomment-472085294

benmoss replied:
if that's the direction we go in for pause then we could do something similar for this image. i'd need to introduce code to pull the image if it's not present.

feiskyer replied:
+1 of the pause container. @PatrickLang Have we included the wincat.exe in the pause image?

benmoss replied:
Good point, we could just use the pause container and `exec` wincat inside it if it's already inside

yujuhong replied:
Packing wincat.exe in the pause image seems like a reasonable workaround. Question: how does this affect the resource usage and the stability of the pause container?

I don't think we need to build everything on the node, even though it may be faster. User/vendors can preload the images if they want. Otherwise, distributing the source to nodes become another issue.

BTW, whatever it's decided here may be used by the containerd integration as well.

yujuhong replied:
Also, the equivalent of pod-level cgroup has not been implemented on Windows, right, @feiskyer @PatrickLang? Just wanted to know how this may affect the resource isolation/accounting in the future.

feiskyer replied:
> Also, the equivalent of pod-level cgroup has not been implemented on Windows, right, 

Right.

neolit123 reviewed (commented):
/priority important-longterm

benmoss commented:
I changed the approach to assume that `wincat.exe` is in the pause image. Running kubelet with `--pod-infra-container-image=gcr.io/cf-london-servces-k8s/wincat/pause`, which I created using code from https://github.com/pivotal-k8s/kubernetes/tree/add-wincat

benmoss commented:
Bumping this, it would be nice to see this make it in.

PatrickLang commented:
@benmoss where did wincat come from?

benmoss commented:
Apologies, I moved the repo to here https://github.com/benmoss/wincat/blob/master/cmd/wincat/wincat.go. Updated the link in the first message as well.

PatrickLang commented:
/lgtm

@yujuhong - can you approve?

feiskyer reviewed (commented):
lgtm.

> I changed the approach to assume that wincat.exe is in the pause image. 

Could you add this to release notes?

yujuhong commented:
/lgtm
/approve

This PR makes it possible for dockershim to support port-forwarding by assuming `wincat` is available in the pause image. I think ideally, this (i.e., the command/binary) should be configurable in the shim, but we don't have any thing in place for dockershim (and probably don't want to add one right now).

Question: @PatrickLang is sig-windows going to maintain the pause image with wincat in it?

benmoss commented:
Just looking this over again after realizing it was actually gonna get merged and had some changes I wanted to make, sorry.

- Look for wincat on the path, rather than at `/wincat.exe`. This gives us more flexibility on where on in the fs it makes sense to actually put it.
- Add a nice error message for when wincat isn't present on the path, since users will likely be seeing this as we settle the pause image standardization problem.

benmoss commented:
/hold cancel

benmoss commented:
/retest

PatrickLang commented:
/milestone v1.15
/lgtm

@yujuhong - yes, https://github.com/kubernetes/kubernetes/pull/75618 is one approach under review to build pause from this repo

madhanrm commented:
@dineshgovindasamy FYI

yujuhong commented on pkg/kubelet/dockershim/docker_streaming_windows.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+func (r *streamingRuntime) portForward(podSandboxID string, port int32, stream io.ReadWriteCloser) error {
+	_, lookupErr := exec.LookPath("wincat.exe")
```
@benmoss this is looking at the path on the host, no? Did you mean to do that?

feiskyer replied:
this is not needed, since we only exec in the sandbox container

benmoss replied:
err, yes my mistake!

yujuhong commented:
/hold

feiskyer reviewed (commented):
looks good now.

/lgtm
/hold cancel

PatrickLang commented:
/retest
